bpo-32230, bpo-20361: Set sys.warnoptions for all command line options

ncoghlan opened the pull request:
Rather than supporting dev mode and the command
line BytesWarning flag directly in the warnings module,
this instead adjusts the initialisation code to add extra
entries to sys.warnoptions when these options are
enabled.

This ensures that dev mode behaves *exactly*
as if `-Wdefault` had been passed on the
command line, including in the way it interacts
with `sys.warnoptions`, and with other command
line flags like `-bb`.

It also ensures that the `-b` and `-bb` options
take precedence of other warnings filter options
like `-Wdefault` or `PYTHONWARNINGS=ignore`.

<!-- issue-number: bpo-32230 -->
https://bugs.python.org/issue32230
<!-- /issue-number -->

<!-- issue-number: bpo-20361 -->
https://bugs.python.org/issue20361
<!-- /issue-number -->

ncoghlan committed:
bpo-32230: Set sys.warnoptions with -X dev

Rather than supporting dev mode directly in
the warnings module, this instead adjusts
the initialisation code to add an extra
'default' entry to sys.warnoptions when
dev mode is enabled.

This ensures that dev mode behaves *exactly*
as if `-Wdefault` had been passed on the
command line, including in the way it interacts
with `sys.warnoptions`, and with other command
line flags like `-bb`.

ncoghlan committed:
Fix typo in NEWS entry

vstinner commented on Lib/test/test_cmd_line.py:
```diff
@@ -541,31 +541,47 @@ def test_xdev(self):
 
         out = self.run_xdev("-c", code)
         self.assertEqual(out,
+                         "default::Warning "
```
default::Warning must come after BytesWarning. See also the discussion on https://bugs.python.org/issue32230

ncoghlan replied:
I'll add a comment pointing out that `Py_BytesWarningFlag` means this filter is pretty redundant when neither `-b` nor `-bb` is passed.

ncoghlan committed:
-X dev should take precedence over PYTHONWARNINGS

As a command line setting, dev mode should take precedence
of the PYTHONWARNINGS environment variable.

ncoghlan committed:
Add explanatory comment about BytesWarning

ncoghlan committed:
Fix Python level checks for a debug build

ncoghlan committed:
Add a comment explaining warnings filter priority order

ncoghlan committed:
Address PR review feedback

- expanded PR to also fix bpo-20361 and have -b & -bb
  take precedence over any other warnings options
- reverted to the short implicit filter list in Py_DEBUG builds
  (this is now even shorter than it was previously, as there
  are *no* implicit filters in debug builds)

ncoghlan committed:
Second NEWS entry, and update What's New

ncoghlan committed:
Add warnings filter test case

ncoghlan committed:
Use the code role to avoid a warning

ncoghlan committed:
Suppress false alarms from 'make suspicious'
